src/components/number-input/number-input.tsx

```diff
@@ -82,24 +83,44 @@ export class NumberInput extends React.Component<NumberInputProps, NumberInputSt
 
     private committed = true;
 
+    private inputRef: HTMLInputElement | null = null;
+
+    private get isUncontrolled(): boolean {
+        return this.props.defaultValue !== undefined;
+    }
+
+    // this accessor reads the value from state
+    // if input is controlled and from the
+    // input itself if it is uncontrolled
+    private get currentValue(): number | undefined {
+        return this.isUncontrolled ?
+            this.inputRef ?
+                this.inputRef.value !== '' ?
+                    Number(this.inputRef.value) :
+                    undefined :
+                this.props.defaultValue :
+            this.state.value;
+    }
```

Comment:
I would like to avoid multiple ternary operators. 
can we do it inside `if` maybe